packages/dds/map/src/directory.ts

```diff
@@ -1537,7 +1537,27 @@ class SubDirectory extends TypedEventEmitter<IDirectoryEvents> implements IDirec
         // This should make the subdirectory structure unreachable so it can be GC'd and won't appear in snapshots
         // Might want to consider cleaning out the structure more exhaustively though?
         const successfullyRemoved = this._subdirectories.delete(subdirName);
-        previousValue?.dispose();
+        this.disposeSubDirectoryTree(previousValue);
         return successfullyRemoved;
     }
+
+    private disposeSubDirectoryTree(directory: IDirectory | undefined) {
+        if (!directory) {
+            return;
+        }
+        const stack: [{node: IDirectory, visited: boolean}] = [{node: directory, visited: false}];
```

Comment:
I'm not sure why you need visited property. It's not a graph, it's a tree, so you should visit each directory once, right?
Or are you afraid of reentrancy, where someone would delete sub-directory from "disposed" event handler?
I general, any changes to data model from handlers should be disallowed. and I'd rather solve this problem centrally, not here. So, I think your code can and should assert that sub-directories you are working with are alive (not disposed).

I'd use recursion instead - a bit easier to read code that way

Reply:
No I was not concerned about reentrancy. Just wanted to fire the dispose events from bottom to top for no string reason. It is just a iterative version of post order traversal. That is why added the visited prop. 
I could use the recursive approach and remove the visited prop. Either way we will use the stack. Implicit in recursion.

Reply:
Got it. That's probably fine, but it would be great to add comments about that. And I'm pretty sure your algorithm does not do what you described - the root (where we started) would get disposed first. Recursive way of doing it would be simpler :)